File: extension/src/operationAuthority.js
'use strict';

const READ_EFFECTS = new Set(['read', 'read_local', 'workspace-read', 'observe', 'host-ui']);
const WRITE_EFFECTS = new Set([
  'write_workspace', 'workspace-write', 'filesystem-write', 'configuration-write', 'clipboard-write',
  'install_tool', 'network', 'run_service', 'process', 'secret_access', 'migration', 'destructive', 'destructive-filesystem'
]);
const EXECUTORS = new Set(['codex-host', 'px-owned-executor']);

function decideAuthority(input = {}) {
  const reasons = [];
  const executor = String(input.executor || '');
  const effects = new Set((input.effects || []).map(String));
  const nonRead = [...effects].some(effect => !READ_EFFECTS.has(effect));
  const workspaceWrite = [...effects].some(effect => [
    'write_workspace', 'workspace-write', 'filesystem-write', 'configuration-write', 'destructive', 'destructive-filesystem'
  ].includes(effect));
  if (!EXECUTORS.has(executor)) reasons.push('executor is not an admitted operation owner');
  if ([...effects].some(effect => !READ_EFFECTS.has(effect) && !WRITE_EFFECTS.has(effect))) reasons.push('request contains unknown effects');
  if (input.observedOnly && nonRead) reasons.push('observation cannot authorize or claim a non-read effect');
  if (nonRead && !input.userApprovalId) reasons.push('non-read effects require current user approval');
  if (nonRead && !input.pxPolicyDecisionId) reasons.push('non-read effects require a PX policy decision');
  if (nonRead && !input.idempotencyKey) reasons.push('non-read effects require an idempotency key');
  if (workspaceWrite && !(input.claimId && input.claimStatus === 'active')) reasons.push('workspace mutation requires an active repository claim');
  if (executor === 'px-owned-executor' && !input.explicitDelegation) reasons.push('PX-owned execution requires explicit delegation');
  const active = [...new Set((input.activeExecutors || []).map(String))];
  if (active.some(owner => !EXECUTORS.has(owner))) reasons.push('active executor set contains an unadmitted owner');
  if (active.some(owner => owner !== executor)) reasons.push('overlapping active executor authority is forbidden');
  if (executor === 'px-owned-executor' && active.includes('codex-host')) reasons.push('PX must not start a nested executor inside the Codex host');
  return Object.freeze({
    allowed: reasons.length === 0,
    executorOwner: reasons.length === 0 ? executor : null,
    reasons: Object.freeze([...new Set(reasons)]),
    requiresUserApproval: nonRead,
    requiresClaim: workspaceWrite
  });
}

function codexHostHandoffDecision({ git, hasWorkspaceClaim = false, requestedEffect = 'workspace-read' } = {}) {
  const gitDecision = git && Array.isArray(git.reasons)
    ? git
    : { allowed: !git?.operation || git.operation === 'none', reasons: git?.operation && git.operation !== 'none' ? [`git-operation:${git.operation}`] : [] };
  const reasons = [...(gitDecision.reasons || [])];
  if (requestedEffect === 'workspace-write' && !hasWorkspaceClaim) reasons.push('workspace-write requires an active repository claim');
  return Object.freeze({
    allowed: gitDecision.allowed !== false && reasons.length === 0,
    mode: 'host-context-handoff',
    executorOwner: 'codex-host',
    extensionExecutes: false,
    requestedEffect,
    reasons: Object.freeze([...new Set(reasons)])
  });
}

module.exports = { decideAuthority, codexHostHandoffDecision, READ_EFFECTS, WRITE_EFFECTS, EXECUTORS };

File: runtime/operation_authority.py
"""Compose operation authority without allowing one control plane to impersonate another.

This module decides *who may own* an operation.  Capability admission, signed effect
grants, policy evidence, and executor-specific validation remain separate gates.
"""

from __future__ import annotations

from dataclasses import dataclass


READ_EFFECTS = frozenset({"read", "read_local", "workspace-read", "observe", "host-ui"})
WRITE_EFFECTS = frozenset(
    {
        "write_workspace",
        "workspace-write",
        "filesystem-write",
        "configuration-write",
        "clipboard-write",
        "install_tool",
        "network",
        "run_service",
        "process",
        "secret_access",
        "migration",
        "destructive",
        "destructive-filesystem",
    }
)
EXECUTORS = frozenset({"codex-host", "px-owned-executor"})


@dataclass(frozen=True)
class AuthorityRequest:
    executor: str
    effects: tuple[str, ...]
    scopes: tuple[str, ...] = ()
    observed_only: bool = False
    user_approval_id: str | None = None
    px_policy_decision_id: str | None = None
    claim_id: str | None = None
    claim_status: str | None = None
    idempotency_key: str | None = None
    explicit_delegation: bool = False
    active_executors: tuple[str, ...] = ()


@dataclass(frozen=True)
class AuthorityDecision:
    allowed: bool
    executor_owner: str | None
    reasons: tuple[str, ...]
    requires_user_approval: bool
    requires_claim: bool


def decide(request: AuthorityRequest) -> AuthorityDecision:
    """Fail closed unless every independent authority supplies its own evidence."""
    reasons: list[str] = []
    effects = set(request.effects)
    non_read = bool(effects - READ_EFFECTS)
    workspace_write = bool(
        effects
        & {
            "write_workspace",
            "workspace-write",
            "filesystem-write",
            "configuration-write",
            "destructive",
            "destructive-filesystem",
        }
    )

    if request.executor not in EXECUTORS:
        reasons.append("executor is not an admitted operation owner")
    unknown_effects = effects - READ_EFFECTS - WRITE_EFFECTS
    if unknown_effects:
        reasons.append("request contains unknown effects")
    if request.observed_only and non_read:
        reasons.append("observation cannot authorize or claim a non-read effect")
    if non_read and not request.user_approval_id:
        reasons.append("non-read effects require current user approval")
    if non_read and not request.px_policy_decision_id:
        reasons.append("non-read effects require a PX policy decision")
    if non_read and not request.idempotency_key:
        reasons.append("non-read effects require an idempotency key")
    if workspace_write and not (
        request.claim_id and request.claim_status == "active"
    ):
        reasons.append("workspace mutation requires an active repository claim")
    if request.executor == "px-owned-executor" and not request.explicit_delegation:
        reasons.append("PX-owned execution requires explicit delegation")

    active = tuple(dict.fromkeys(request.active_executors))
    if any(owner not in EXECUTORS for owner in active):
        reasons.append("active executor set contains an unadmitted owner")
    other_owners = {owner for owner in active if owner != request.executor}
    if other_owners:
        reasons.append("overlapping active executor authority is forbidden")
    if request.executor == "px-owned-executor" and "codex-host" in active:
        reasons.append("PX must not start a nested executor inside the Codex host")

    return AuthorityDecision(
        allowed=not reasons,
        executor_owner=request.executor if not reasons else None,
        reasons=tuple(dict.fromkeys(reasons)),
        requires_user_approval=non_read,
        requires_claim=workspace_write,
    )


def authority_roles() -> dict[str, dict[str, object]]:
    """Return the stable role boundary used by diagnostics and user interfaces."""
    return {
        "codex-host": {
            "may_execute": True,
            "owns_user_approval_surface": True,
            "may_issue_px_policy": False,
            "may_issue_repository_claim": False,
        },
        "px-control-plane": {
            "may_execute": False,
            "owns_user_approval_surface": False,
            "may_issue_px_policy": True,
            "may_issue_repository_claim": False,
        },
        "repository-claim": {
            "may_execute": False,
            "owns_user_approval_surface": False,
            "may_issue_px_policy": False,
            "may_issue_repository_claim": True,
        },
        "extension": {
            "may_execute": False,
            "owns_user_approval_surface": False,
            "presentation_and_observation_only": True,
        },
    }
